pkg/statsd/dispatcher.go

```diff
@@ -48,30 +52,61 @@ func NewMetricDispatcher(numWorkers int, perWorkerBufferSize int, af AggregatorF
 	}
 }
 
-// Run runs the MetricDispatcher.
-func (d *MetricDispatcher) Run(ctx context.Context, done gostatsd.Done) {
-	defer done()
-	var wg sync.WaitGroup
-	wg.Add(d.numWorkers)
-	for _, worker := range d.workers {
-		w := worker // Make a copy of the loop variable! https://github.com/golang/go/wiki/CommonMistakes
-		go w.work(wg.Done)
-	}
-	defer func() {
+// RunAsync runs the MetricDispatcher in a goroutine
+func (d *MetricDispatcher) RunAsync(ctx context.Context, done gostatsd.Done) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.running = true
+	go func() {
+		defer done()
+
+		var wg sync.WaitGroup
+		wg.Add(d.numWorkers)
 		for _, worker := range d.workers {
-			close(worker.metricsQueue) // Close channel to terminate worker
+			w := worker // Make a copy of the loop variable! https://github.com/golang/go/wiki/CommonMistakes
+			go w.work(wg.Done)
 		}
-		wg.Wait() // Wait for all workers to finish
+
+		defer func() {
+			// Once this is indicated as not running, DispatchMetric will no longer
+			// try to send to the channel, so it is safe to close.
+			d.lock.Lock()
+			d.running = false
+			d.lock.Unlock()
+
+			for _, worker := range d.workers {
+				close(worker.metricsQueue) // Close channel to terminate worker
+			}
+			wg.Wait() // Wait for all workers to finish
+		}()
+
+		// Work until asked to stop
+		<-ctx.Done()
 	}()
+}
 
-	// Work until asked to stop
-	<-ctx.Done()
+func (d *MetricDispatcher) runMetrics(ctx context.Context, statser statser.Statser) {
+	for _, worker := range d.workers {
+		worker.runMetrics(ctx, statser)
+	}
+	d.Process(ctx, func(aggrId uint16, aggr Aggregator) {
+		tag := fmt.Sprintf("aggregator_id:%d", aggrId)
+		aggr.TrackMetrics(statser.WithTags(gostatsd.Tags{tag}))
+	})
 }
 
 // DispatchMetric dispatches metric to a corresponding Aggregator.
 func (d *MetricDispatcher) DispatchMetric(ctx context.Context, m *gostatsd.Metric) error {
 	hash := adler32.Checksum([]byte(m.Name))
 	w := d.workers[uint16(hash%uint32(d.numWorkers))]
+
+	// Protects both the read of d.running, and that the channel remains open
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	if !d.running {
+		return nil
+	}
+
```

Comment:
It does, which I'm definitely not thrilled about.  RWMutex was the best I could come up with.  If there's a better way, I'm open to it.  The problem I had is the CSW tries to submit to a closing channel, which is a data race.

I originally had an explicit stop on the CSWs, but then I need to make sure I have a backdoory path to always notify CSWs to stop in future.  I'd rather they continue to be able to "submit" but have it drop.